src/wallet/wallet.go

```diff
@@ -28,3 +36,387 @@ func NewWalletFilename() string {
+		if e.Address == a {
+			return e, true
+		}
+	}
+	return Entry{}, false
+}
+
+// AddEntry adds new entry
+func (wlt *Wallet) AddEntry(entry Entry) error {
+	// dup check
+	for _, e := range wlt.Entries {
+		if e.Address == entry.Address {
+			return errors.New("duplicate address entry")
+		}
+	}
+
+	wlt.Entries = append(wlt.Entries, entry)
+	return nil
+}
+
+// Save persists wallet to disk
+func (wlt *Wallet) Save(dir string) error {
+	r := NewReadableWallet(*wlt)
+	return r.Save(filepath.Join(dir, wlt.GetFilename()))
+}
+
+// Load loads wallets from given dir
+func (wlt *Wallet) Load(dir string) error {
+	r := &ReadableWallet{}
+	if err := r.Load(filepath.Join(dir, wlt.GetFilename())); err != nil {
+		return err
+	}
+	r.Meta["filename"] = wlt.GetFilename()
+	*wlt = NewWalletFromReadable(r)
+	return nil
+}
+
+// Validator validate if the wallet be able to create spending transaction
+type Validator interface {
+	// checks if any of the given addresses has unconfirmed spending transactions
+	HasUnconfirmedSpendTx(addr []cipher.Address) (bool, error)
+}
+
+// CreateAndSignTransaction Creates a Transaction
+// spending coins and hours from wallet
+func (wlt *Wallet) CreateAndSignTransaction(
+	vld Validator,
+	unspent blockdb.UnspentGetter,
+	headTime uint64,
+	amt Balance,
+	dest cipher.Address) (*coin.Transaction, error) {
+
+	addrs := wlt.GetAddresses()
+	ok, err := vld.HasUnconfirmedSpendTx(addrs)
+	if err != nil {
+		return nil, fmt.Errorf("Checking unconfirmed spending failed: %v", err)
+	}
+
+	if ok {
+		return nil, errors.New("Please spend after your pending transaction is confirmed")
+	}
+
+	txn := coin.Transaction{}
+	auxs := unspent.GetUnspentsOfAddrs(wlt.GetAddresses())
+
+	// Determine which unspents to spend
+	spends, err := createSpends(headTime, auxs.Flatten(), amt)
+	if err != nil {
+		return nil, err
+	}
+
+	// Add these unspents as tx inputs
+	toSign := make([]cipher.SecKey, len(spends))
+	spending := Balance{Coins: 0, Hours: 0}
+	for i, au := range spends {
+		entry, exists := wlt.GetEntry(au.Body.Address)
+		if !exists {
+			return nil, fmt.Errorf("Address:%v does not exist in wallet:%v", au.Body.Address, wlt.GetID())
+		}
+
+		txn.PushInput(au.Hash())
+		toSign[i] = entry.Secret
+		spending.Coins += au.Body.Coins
+		spending.Hours += au.CoinHours(headTime)
+	}
+
+	//keep 1/4th of hours as change
+	//send half to each address
+	var changeHours = uint64(spending.Hours / 4)
+
+	if amt.Coins == spending.Coins {
+		txn.PushOutput(dest, amt.Coins, changeHours/2)
+		txn.SignInputs(toSign)
+		txn.UpdateHeader()
+		return &txn, nil
+	}
+
+	change := NewBalance(spending.Coins-amt.Coins, changeHours/2)
+	// TODO -- send change to a new address
+	changeAddr := spends[0].Body.Address
+
+	//create transaction
+	txn.PushOutput(changeAddr, change.Coins, change.Hours)
+	txn.PushOutput(dest, amt.Coins, changeHours/2)
+	txn.SignInputs(toSign)
+	txn.UpdateHeader()
+	return &txn, nil
+}
+
+func createSpends(headTime uint64, uxa coin.UxArray,
```

Comment:
this function does not stop appending `UxOut` to `spending` once it has enough to satisfy amount?  all it seems to do is filter out invalid `UxOut` and return error if sum is less than balance